ARTEMIS-3596 Pass class loader to ServiceLoader.load to fix OSGi issues

ryeats opened the pull request:
This resolves an issue i was seeing when running artemis in karaf where if the artemis-server-osgi bundle got reloaded during the resolution process I would see this error prevent artemis from starting.
org.apache.activemq.artemis.spi.core.remoting.ssl.SSLContextFactory: org.apache.activemq.artemis.core.remoting.impl.ssl.DefaultSSLContextFactory not a subtype

clebertsuconic commented on artemis-core-client/src/main/java/org/apache/activemq/artemis/spi/core/remoting/ssl/OpenSSLContextFactoryProvider.java:
```diff
@@ -25,7 +25,7 @@
    private static final OpenSSLContextFactory FACTORY;
    static {
       OpenSSLContextFactory factoryWithHighestPrio = null;
-      for (OpenSSLContextFactory factory : ServiceLoader.load(OpenSSLContextFactory.class)) {
```
Can't you use ThreadContextClass Loader on all the tests here?

it would be nice to have a test.

ryeats replied:
OSGi in certain conditions has issues with ServiceLoader.load(<class>.class) which uses Thread.currentThread().getContextClassLoader() because that classloader is not always the same as the one that is available to the bundle throughout its life. This change should be transparent for any environment that doesn’t juggle class-loaders like OSGi since either way would end up returning the same class-loader.

As far as writing a test its difficult since the conditions i was seeing errors are hard to recreate see the JIRA for details.

ryeats replied:
OSGi in certain conditions has issues with ServiceLoader.load(<class>.class) which uses Thread.currentThread().getContextClassLoader() because that classloader is not always the same as the one that is available to the bundle throughout its life. This change should be transparent for any environment that doesn’t juggle class-loaders like OSGi since either way would end up returning the same class-loader.

 As far as writing a test its difficult since the conditions i was seeing errors are hard to recreate see the JIRA for details.

ryeats commented:
I changed it the calls to use the calling classes classloader since that is what makes more sense and was done elsewhere in the code.

clebertsuconic commented:
@ryeats fair enough... I'm running a full CI test suite on it.. it takes 3 hours to perform and there are 2 runs ahead of this. I will let you know how it goes tomorrow. if all good I will merge unless someone else sees an issue I don't see.

clebertsuconic commented:
well we have a few failures because of this:

org.apache.activemq.artemis.tests.integration.karaf.ArtemisFeatureTest.test
org.apache.activemq.artemis.tests.integration.management.SSLSecurityNotificationTest.testSECURITY_AUTHENTICATION_VIOLATION
org.apache.activemq.artemis.tests.integration.management.SSLSecurityNotificationTest.testCONSUMER_CREATED
org.apache.activemq.artemis.tests.integration.security.SecurityTest.testJAASSecurityManagerAuthenticationWithRegexpsWantClientAuth
org.apache.activemq.artemis.tests.integration.security.SecurityTest.testJAASSecurityManagerAuthorizationPositiveWithCertsWantClientAuth
org.apache.activemq.artemis.tests.integration.security.SecurityTest.testJAASSecurityManagerAuthorizationPositiveWithCerts
org.apache.activemq.artemis.tests.integration.security.SecurityTest.testJAASSecurityManagerAuthenticationWithCerts
org.apache.activemq.artemis.tests.integration.security.SecurityTest.testJAASSecurityManagerAuthenticationWithBadClientCert
org.apache.activemq.artemis.tests.integration.security.SecurityTest.testJAASSecurityManagerAuthenticationWithCertsWantClientAuth
org.apache.activemq.artemis.tests.integration.security.SecurityTest.testJAASSecurityManagerOpenWireNegative
org.apache.activemq.artemis.tests.integration.security.SecurityTest.testJAASSecurityManagerAuthenticationWithCertsAndOpenWire
org.apache.activemq.artemis.tests.integration.security.SecurityTest.testJAASSecurityManagerAuthenticationWithRegexps
org.apache.activemq.artemis.tests.integration.security.SecurityTest.testJAASSecurityManagerAuthorizationNegativeWithCerts
org.apache.activemq.artemis.tests.integration.ssl.CoreClientOverOneWaySSLTest.testOneWaySSL[storeType=JCEKS]
org.apache.activemq.artemis.tests.integration.ssl.CoreClientOverOneWaySSLTest.testOneWaySSLUsingDefaultSslContext[storeType=JCEKS]
org.apache.activemq.artemis.tests.integration.ssl.CoreClientOverOneWaySSLTest.testOneWaySSLwithURLandMaskedPasswordProperty[storeType=JCEKS]
org.apache.activemq.artemis.tests.integration.ssl.CoreClientOverOneWaySSLTest.testOneWaySSLVerifyHost[storeType=JCEKS]
org.apache.activemq.artemis.tests.integration.ssl.CoreClientOverOneWaySSLTest.testOneWaySSLReloaded[storeType=JCEKS]
org.apache.activemq.artemis.tests.integration.ssl.CoreClientOverOneWaySSLTest.testOneWaySSLwithSNIOnlyOnTheBroker[storeType=JCEKS]
org.apache.activemq.artemis.tests.integration.ssl.CoreClientOverOneWaySSLTest.testOneWaySSLwithSNIOnlyOnTheClient[storeType=JCEKS]
org.apache.activemq.artemis.tests.integration.ssl.CoreClientOverOneWaySSLTest.testOneWaySSLwithSNI[storeType=JCEKS]
org.apache.activemq.artemis.tests.integration.ssl.CoreClientOverOneWaySSLTest.testOneWaySSLwithURL[storeType=JCEKS]
org.apache.activemq.artemis.tests.integration.ssl.CoreClientOverOneWaySSLTest.testOneWaySSLwithURLandMaskedPasswordENCSyntax[storeType=JCEKS]
org.apache.activemq.artemis.tests.integration.ssl.CoreClientOverOneWaySSLTest.testOneWaySSLWithGoodClientProtocol[storeType=JCEKS]
org.apache.activemq.artemis.tests.integration.ssl.CoreClientOverOneWaySSLTest.testOneWaySSLWithGoodServerProtocol[storeType=JCEKS]
org.apache.activemq.artemis.tests.integration.ssl.CoreClientOverOneWaySSLTest.testOneWaySSLwithTrustManagerPlugin[storeType=JCEKS]
org.apache.activemq.artemis.tests.integration.ssl.CoreClientOverOneWaySSLTest.testOneWaySSLWithGoodClientProtocol[storeType=JKS]
org.apache.activemq.artemis.tests.integration.ssl.CoreClientOverOneWaySSLTest.testOneWaySSLWithGoodServerProtocol[storeType=JKS]
org.apache.activemq.artemis.tests.integration.ssl.CoreClientOverOneWaySSLTest.testOneWaySSLWithGoodClientProtocol[storeType=PKCS12]
org.apache.activemq.artemis.tests.integration.ssl.CoreClientOverOneWaySSLTest.testOneWaySSLWithGoodServerProtocol[storeType=PKCS12]
org.apache.activemq.artemis.tests.integration.ssl.CoreClientOverTwoWayOpenSSLServerTest.testTwoWaySSLWithoutClientKeyStore[storeType=JCEKS]
org.apache.activemq.artemis.tests.integration.ssl.CoreClientOverTwoWayOpenSSLServerTest.testTwoWaySSLVerifyClientHostNegative[storeType=JCEKS]
org.apache.activemq.artemis.tests.integration.ssl.CoreClientOverTwoWayOpenSSLServerTest.testTwoWaySSLWithoutClientKeyStore[storeType=JKS]
org.apache.activemq.artemis.tests.integration.ssl.CoreClientOverTwoWayOpenSSLServerTest.testTwoWaySSLVerifyClientHostNegative[storeType=JKS]
org.apache.activemq.artemis.tests.integration.ssl.CoreClientOverTwoWayOpenSSLTest.testTwoWaySSLWithoutClientKeyStore[storeType=JCEKS]
org.apache.activemq.artemis.tests.integration.ssl.CoreClientOverTwoWayOpenSSLTest.testTwoWaySSLWithoutClientKeyStore[storeType=JKS]
org.apache.activemq.artemis.tests.integration.ssl.CoreClientOverTwoWaySSLTest.testTwoWaySSLVerifyClientTrustAllTrue[storeType=JCEKS]
org.apache.activemq.artemis.tests.integration.ssl.CoreClientOverTwoWaySSLTest.testTwoWaySSLVerifyClientHost[storeType=JCEKS]
org.apache.activemq.artemis.tests.integration.ssl.CoreClientOverTwoWaySSLTest.testTwoWaySSL[storeType=JCEKS]
org.apache.activemq.artemis.tests.integration.ssl.CoreClientOverTwoWaySSLTest.testTwoWaySSLVerifyClientTrustAllTrueByURI[storeType=JCEKS]
org.apache.activemq.artemis.tests.integration.ssl.CoreClientOverTwoWaySSLTest.testTwoWaySSLVerifyClientTrustAllTrue[storeType=JKS]
org.apache.activemq.artemis.tests.integration.ssl.CoreClientOverTwoWaySSLTest.testTwoWaySSL[storeType=JKS]
org.apache.activemq.artemis.tests.integration.ssl.CoreClientOverTwoWaySSLTest.testTwoWaySSLWithoutClientKeyStore[storeType=JKS]
org.apache.activemq.artemis.tests.integration.ssl.CoreClientOverTwoWaySSLTest.testTwoWaySSLVerifyClientTrustAllTrueByURI[storeType=JKS]
org.apache.activemq.artemis.tests.integration.ssl.CoreClientOverTwoWaySSLTest.testTwoWaySSLWithoutClientKeyStore[storeType=PKCS12]
org.apache.activemq.artemis.tests.integration.ssl.DualAuthenticationTest.testDualAuthentication

ryeats commented:
Okay ill figure out how to run those tests and see whats going on.

ryeats commented:
I am having trouble narrowing down the issue, I don't believe it was these changes but I was struggling to figure out what was causing the test errors. I was seeing failures with and without the changes from this PR and then i rebased and it seems to be passing now.

clebertsuconic commented:
These tests pass on a daily basis without any issues without these changes

clebertsuconic commented:
I was talking about the CI tests.. not sure if that's what you meant

ryeats commented:
Yes, I understand that is why it took me so long to respond initially. While I was trying to figure out what was going on I was seeing issues but possibly not the issues you saw because these didn't seem to be because of my changes. Then I rebased and my problem now is they are not failing anymore and I didn't do anything to fix it. When I debug through the code I changed with or without my change it returns the same class from ServiceLoad.load each time so i don't have a lot to go on. I was hoping you had an idea since i am going to be going back through recent commits to see which one fixed the issue we were seeing otherwise.

ryeats commented:
Apologies I just realized I was rebased off of master instead of main when i initially put up the PR so there was actually quite a lot of changes that could have fixed the issue.

clebertsuconic commented:
@ryeats you mean the failures were just because you were behind main?

ryeats commented:
Yes, I was building off of a commit that had itest errors back from march

clebertsuconic commented:
@ryeats I see... I will merge and I will trust you ;) 

I will run the test suite again.. and If I see weird failures I will let you know..
